Relocate load-balancing configuration options to loadbalancer.Config

joamaki opened the pull request:
This moves a good chunk of the load-balancing configuration options from `option.DaemonConfig` to `loadbalancer.Config`.

The only remaining config option that the control-plane uses is the `KubeProxyReplacement` which is a bit harder to move so leaving that to a separate PR.

Bunch of options that only affect datapath (NodePortMode*, NodePortAcceleration*, etc.) were also left untouched. Unclear whether we should keep them in `loadbalancer.Config` or whether they'd belong into e.g. `datapath.Config`.

joamaki committed:
loadbalancer: Move LB maps related config options to loadbalancer package

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
loadbalancer: Move NodePortRange to loadbalancer.Config

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
loadbalancer: Move LBAlgorithm to loadbalancer.Config

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
loadbalancer: Move ExternalClusterIP to loadbalancer.Config

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
loadbalancer: Move LoadBalancerAlgorithmAnnotation to Config

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
loadbalancer: Move EnableHealthCheckNodePort to Config

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
Documentation: update cmdref after flags moved

Bunch of flags were moved from DaemonConfig to hive configs and these
now appear in listed flags for the "hive" etc. commands.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

sayboras reviewed (approved):
Looks good for the file owned by cilium/proxy group (e.g. pkg/ciliumenvoyconfig/exp_test.go) 👍

joamaki committed:
loadbalancer/reflectors: Only complete when all observables complete

The stress test (stress.sh) hit the following panic:

  panic: *resource.resource[*github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1.Service].Events() called from
     .../resource/resource.go:376 has a broken event handler that did not call Done() before event was garbage collected

This was caused by the [joinObservables] immediately stopping to emit events when one of the observables
completes which caused missing a call to Done().

Fix this by changing [joinObservables] to complete only once all source observables complete
and by calling ev.Done(nil) before joinObservables processes it. This way there isn't a way
in which we would not call Done().

Verified by running 5000 stress test iterations (instead of the usual 500 that stress.sh does).

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented on pkg/loadbalancer/redirectpolicy/testdata/address.txtar:
```diff
@@ -129,7 +129,7 @@ BE: ID=1 ADDR=10.244.2.1:80/TCP STATE=active
 REV: ID=1 ADDR=169.254.169.254:8080
 REV: ID=2 ADDR=[2001::1]:8080
-SVC: ID=1 ADDR=169.254.169.254:8080/TCP SLOT=0 LBALG=undef AFFTimeout=0 COUNT=1 QCOUNT=0 FLAGS=LocalRedirect
+SVC: ID=1 ADDR=169.254.169.254:8080/TCP SLOT=0 LBALG=random AFFTimeout=0 COUNT=1 QCOUNT=0 FLAGS=LocalRedirect
```
FIXME: This is not right. We should not annotate unless the annotation is actually set for this service and that the algorithm can be used for this service (e.g. maglev usually not used for ClusterIP)

joamaki committed:
loadbalancer: Only set algorithm annotation when requested

The algorithm in the service entry should only be set if annotations
are enabled and the entry has the annotation. This fixes the issue that
e.g. ClusterIP services were annotated as maglev even though by default
(without ExternalClusterIP) we don't provision maglev tables for them.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

ti-mo commented on pkg/datapath/linux/config/config.go:
```diff
@@ -82,6 +84,7 @@ func NewHeaderfileWriter(p WriterParams) (datapath.ConfigWriter, error) {
 		log:                p.Log,
 		sysctl:             p.Sysctl,
 		maglev:             p.Maglev,
+		lbConfig:           p.LBConfig,
```
More of a high-level design/org question: would it be possible to plumb this through (lb/maglev/sysctl) LocalNodeConfiguration instead? Asking because most of WriteNodeConfig will be migrated to [nodeConfig() in pkg/datapath/loader](https://github.com/cilium/cilium/blob/61a7033a3b81080a046db48122cc8a33c9e3a8bb/pkg/datapath/loader/config.go#L12) and I want to keep the porting work as straightforward as possible. 

`nodeConfig()` already receives a LNC, and I don't think making all loader entrypoints take feature config objects is a good idea. (Maybe we should also freeze additions to `HeaderfileWriter` like you've suggested here for this reason?)

joamaki replied:
Sounds like a good idea. I’ll move it (and maybe other deps too) into LNC.

EDIT: Not entirely sure about sysctl and maglev as they aren't configuration so I'll leave changing those out from this PR. `nodeAddressing` for sure is configuration and should come via LNC (the whole NodeAddressing interface should be removed).

ti-mo replied:
Thanks for looking into it! Re: sysctl: makes sense, looks like it's a query interface into actual sysctl data.

Re: maglev: I think WriterParams (or, LNC actually) should embed a `maglev.Config`, not a `maglev.Maglev`. WriteNodeConfig accesses `h.maglev.Config.MaglevTableSize`, `h.maglev.SeedJhash0` and `h.maglev.SeedJhash1`, so I'd argue SeedJhash0/1 could be refactored to be a part of `maglev.Config` instead. Could that work?

joamaki replied:
That could work yeah! Would need to split `maglev.Config` into the "user facing immutable config" and the actual `Config` that has the seeds computed, but that's straightforward. Will do another PR for this.

joamaki replied:
https://github.com/cilium/cilium/pull/39258

giorio94 reviewed (approved):
/lgtm for the clustermesh changes.

derailed reviewed (approved):
@joamaki Nice work!

brb reviewed (approved):
:heart:
